File: 2023/rust/src/day_01.rs
// https://adventofcode.com/2023/day/1
use std::fs::File;
use std::io::{self, Read};
use std::path::Path;

const TEXT_DIGITS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn read_file_contents(input_fname: impl AsRef<Path>) -> io::Result<String> {
    let file_path = input_fname.as_ref();
    let mut input = File::open(file_path)?;
    let mut contents = String::new();
    input.read_to_string(&mut contents)?;

    Ok(contents)
}

fn process_line(line: &str, with_text: bool) -> (i32, i32) {
    let mut first_digit = -1;
    let mut last_digit = -1;

    for (i, ch) in line.chars().enumerate() {
        if ch.is_digit(10) {
            last_digit = ch.to_digit(10).unwrap() as i32;
        } else {
            if !with_text {
                continue;
            }
            for (index, text_digit) in TEXT_DIGITS.iter().enumerate() {
                if line[i..].starts_with(text_digit) {
                    last_digit = index as i32 + 1;
                }
            }
        }
        if first_digit == -1 {
            first_digit = last_digit;
        }
    }

    (first_digit, last_digit)
}

pub fn solve_day_01_part_1(input_fname: impl AsRef<Path>) -> i32 {
    let file_contents = read_file_contents(input_fname).expect("Failed to read file");

    let mut result = 0;
    for line in file_contents.lines() {
        let (first_digit, last_digit) = process_line(line, false);
        result += first_digit * 10 + last_digit;
    }
    result
}

pub fn solve_day_01_part_2(input_fname: impl AsRef<Path>) -> i32 {
    let file_contents = read_file_contents(input_fname).expect("Failed to read file");
    let mut result = 0;
    for line in file_contents.lines() {
        let (first_digit, last_digit) = process_line(line, true);
        result += first_digit * 10 + last_digit;
    }

    result
}

File: 2023/python/day_01.py
# https://adventofcode.com/2023/day/1

INPUT_FILE: str = "2023/data/day_01.txt"

DIGITS_AS_TEXT = (
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
)


def get_input(input_file: str) -> list[str]:
    with open(input_file, "r") as f:
        return [x.strip() for x in f.readlines()]


def solve_day_01_part_1(input_file: str = INPUT_FILE):
    lines = get_input(input_file)

    result: int = 0
    for line in lines:
        digits = [int(x) for x in line if x in "0123456789"]
        result += digits[0] * 10 + digits[-1]

    return result


def solve_day_01_part_2(input_file: str = INPUT_FILE):
    lines = get_input(input_file)

    res = 0
    for line in lines:
        digits = []
        for i in range(len(line)):
            if line[i] in "0123456789":
                digits.append(int(line[i]))
            else:
                for j, wd in enumerate(DIGITS_AS_TEXT):
                    if line[i:].startswith(wd):
                        digits.append(j + 1)
        res += digits[0] * 10 + digits[-1]

    return res


if __name__ == "__main__":
    print(f"Solution for Part 1: {solve_day_01_part_1(INPUT_FILE)}")
    print(f"Solution for Part 2: {solve_day_01_part_2(INPUT_FILE)}")
